rust/agama-server/src/software/web.rs

```diff
@@ -0,0 +1,242 @@
+//! This module implements the web API for the software module.
+//!
+//! The module offers two public functions:
+//!
+//! * `software_service` which returns the Axum service.
+//! * `software_stream` which offers an stream that emits the software events coming from D-Bus.
+
+use crate::{error::Error, web::Event};
+use agama_lib::{
+    error::ServiceError,
+    product::{Product, ProductClient},
+    software::{
+        proxies::{Software1Proxy, SoftwareProductProxy},
+        Pattern, SelectedBy, SoftwareClient,
+    },
+};
+use axum::{
+    extract::State,
+    http::StatusCode,
+    response::{IntoResponse, Response},
+    routing::{get, post, put},
+    Json, Router,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::collections::HashMap;
+use thiserror::Error;
+use tokio_stream::{Stream, StreamExt};
+
+#[derive(Clone)]
+struct SoftwareState<'a> {
+    product: ProductClient<'a>,
+    software: SoftwareClient<'a>,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+struct SoftwareConfig {
+    patterns: Option<Vec<String>>,
+    product: Option<String>,
+}
+
+#[derive(Error, Debug)]
+pub enum SoftwareError {
+    #[error("Service error: {0}")]
```

Comment:
I am not sure if we should not specify which service like `Software Service Error: {0}`